### DISCUSSION TITLE
【23年7月更】M401A/CM311等盒子Armbian 系统 RTL8761BTV 蓝牙的最正确食用方式

### DISCUSSION THREAD

**Discussion by xcray** [ophub/amlogic-s9xxx-armbian#1546] [Amlogic]
原帖：https://bbs.hassbian.com/thread-18366-1-1.html
有时候github登不上，需要讨论的话可以移步。

很惭愧半年多没跟进o大的动态，最近重刷m401a盒子发现o的镜像比去年有了非常大的优化，安装变得非常简捷，狂赞！

但是蓝牙还是需要手动调整才能正常使用，看到几个朋友发帖，似乎还是没理顺。特把基于最新镜像的个人认为迄今最正确的蓝牙启用方法整理如下，供大家参考：

### 方法步骤：
0. 系统安装
我这次用的是Armbian_23.08.0_amlogic_s905l3a_bookworm_6.1.36_server_2023.07.01，修改uEnv指定m401a.dtb，其他未作任何修改，安装到emmc一次成功，非常顺利！

然后apt全面更新、内核更新到最新版，armbian-config里安装蓝牙组件。

1. 反编译dtb、修改dts、在serial@24000部分增加如下bluetooth内容并重新编译为dtb，替换掉原来的dtb：
`       bluetooth {
                compatible = "realtek,rtl8723bs-bt";
        };
`
2. 将之前用的2m波特率的config文件复制到/lib/firmware/rtl_bt/rtl8761b_config.bin。系统原来自带的config文件不能用。
3. 在/lib/systemd/system/bluetooth.service的[Service]部分加一行：
`ExecStopPost=/usr/bin/env gpioset 0 82=0`
4. 重启、测试

### 讨论解释：
1. 原来我编译过rtk_hciattach，我看现在o大把它放进了镜像里。这是一个用户空间的工具，在用户空间完成蓝牙电路的附着。这种方式比较陈旧！
2. 我上面说的修改dtb的方法，是在内核空间直接驱动（serdev），是一种更为现代化的方法，当然也更推荐。
参考：https://github.com/ophub/amlogic-s9xxx-armbian/issues/639#issuecomment-1289901906，[lovelylain](https://github.com/lovelylain)的解释还是很到位的！
实际用下来，这种方式确实比原来的要好！简洁、启动快不说，稳定性也有明显提升——
用2m速率的config文件ha蓝牙集成连续多天运行也不再出现传感器停更的问题了！
而之前顶多能坚持几个小时。
为什么用“rtl8723bs-bt”而不用8761？这是因为内核里的驱动程序，具体我也闹不太明白，懂的都懂哈。
4. 上面第3步的含义是在系统重启过程中重置蓝牙电路，不重置的话重启后是无法使用蓝牙的。
5. 还有个没搞懂的问题：按说rfkill是可以自动完成蓝牙电路重置的任务的，只是搞不懂具体怎么弄。
6. 其他型号的蓝牙应该也可以参考这个思路。
7. 建议o大将这些内容合并到未来的镜像中，这样刷机后蓝牙直接就是活的。

**Comment by maxmib** [comment]
能否分享一下最新编译的dtb文件，我使用你之前在帖子里发的dtb文件蓝牙能够正常使用，但是如果接了移动硬盘会大概率导致开机的时候蓝牙和硬盘都无法加载

**Reply by IOsetting** [reply]
是不是供电跟不上? 移动硬盘换成带外置供电的试试

**Reply by maxmib** [reply]
不是供电的问题
有类似的issue  #982

**Comment by IOsetting** [comment]
这个盒子在 CoreELEC/EmuELEC下的蓝牙有人启用成功的么? 内核是4.9的, 和Armbian差别比较大.

**Comment by WalterDSU** [comment]
有人知道这个有添加到最新的armbian里面了吗？

**Reply by a18ccms** [reply]
没有

**Reply by superyaya9** [reply]
太可惜了，2025年了。不知什么时候能更新进去

**Comment by buer23333** [comment]
大佬，可以出一个详细的教程吗？对于小白来说有点难以理解

**Reply by buer23333** [reply]
我利用dtc -I dtb -O dts -o meson-g12a-s905l3a-m401a.dts meson-g12a-s905l3a-m401a.dtb命令，但是出现以下报错：
meson-g12a-s905l3a-m401a.dts: Warning (unit_address_vs_reg): /soc/bus@ff600000/bus@34400/pinctrl@40: node has a unit name, but no reg or ranges property
meson-g12a-s905l3a-m401a.dts: Warning (unit_address_vs_reg): /soc/bus@ff800000/sys-ctrl@0/pinctrl@14: node has a unit name, but no reg or ranges property
meson-g12a-s905l3a-m401a.dts: Warning (simple_bus_reg): /soc/bus@ff600000/bus@34400/pinctrl@40: missing or empty reg/ranges property
meson-g12a-s905l3a-m401a.dts: Warning (gpios_property): /__symbols__:gpio: property size (47) is invalid, expected multiple of 4
meson-g12a-s905l3a-m401a.dts: Warning (interrupt_provider): /soc/bus@ffd00000/interrupt-controller@f080: Missing #address-cells in interrupt provider

**Comment by thdaemon** [comment]
最近入了一个m401a，遇到了同样的问题，感谢提供信息。看了一下驱动代码，只需要给 dts 加 `enable-gpios` 就可以了，改 bluetooth.service 不太好。

https://github.com/unifreq/linux-6.12.y

```
diff --git a/arch/arm64/boot/dts/amlogic/meson-g12a-s905l3a-m401a.dts b/arch/arm64/boot/dts/amlogic/meson-g12a-s905l3a-m401a.dts
index a538d0246..f464233dc 100644
--- a/arch/arm64/boot/dts/amlogic/meson-g12a-s905l3a-m401a.dts
+++ b/arch/arm64/boot/dts/amlogic/meson-g12a-s905l3a-m401a.dts
@@ -61,6 +61,10 @@ &uart_A {
        pinctrl-0 = <&uart_a_pins>, <&uart_a_cts_rts_pins>;
        uart-has-rtscts;
        pinctrl-names = "default";
+       bluetooth {
+               compatible = "realtek,rtl8723bs-bt";
+               enable-gpios = <&gpio GPIOX_17 GPIO_ACTIVE_HIGH>;
+       };
 };
 
 &mali {
```

编译：

```
make ARCH=arm64 dtbs
```

不过我看驱动还会读一个 `device-wake-gpios` 的，不知道是干嘛的，是不是设备休眠唤醒需要用这个？有没有人知道？

感觉可以看看原厂 android dtb 反编译，不知道是不是一样的

**Comment by steven-ye** [comment]
经测试，CM311-1a的dts加下面的，用dtc编译后就可以了。在debian上编译的dtb, 在openwrt上也测试通过，说明可行
        bluetooth {
          compatible = "realtek,rtl8723bs-bt";
		      enable-gpios = <0x35 0x52 0x00>;
		      reset-gpios = <0x35 0x52 0x01>;
	      };
